File: src/storage/rclone.ts
import { mkdir } from "node:fs/promises"
import { existsSync } from "node:fs"
import { join } from "node:path"
import { homedir } from "node:os"
import { execSync } from "node:child_process"
import type { StorageBackend } from "./types"

const BASE_OPTS = {
  encoding: "utf8" as const,
  stdio: ["pipe", "pipe", "pipe"] as ["pipe", "pipe", "pipe"],
}

function cleanArg(a: string): string {
  // Just wrap in quotes, escape inner quotes
  return `"${a.replace(/\"/g, '\\"')}"`
}

/** Find the rclone config file. Hermes changes $HOME so the default
 *  ~/.config/rclone/rclone.conf may point to the wrong place. */
function findConfig(): string {
  const candidates = [
    join(homedir(), ".config", "rclone", "rclone.conf"),
    "/home/keraki/.config/rclone/rclone.conf",
    "/root/.config/rclone/rclone.conf",
  ]
  for (const path of candidates) {
    if (existsSync(path)) return path
  }
  return join(homedir(), ".config", "rclone", "rclone.conf")
}

const RCLONE_CONFIG = findConfig()

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms))
}

function rclone(args: string[], timeoutMs?: number): { stdout: string; stderr: string; exitCode: number } {
  const joined = args.map(cleanArg).join(" ")
  const cmd = `rclone --config "${RCLONE_CONFIG}" ${joined}`
  const opts = timeoutMs !== undefined ? { ...BASE_OPTS, timeout: timeoutMs } : BASE_OPTS
  try {
    const stdout = execSync(cmd, opts)
    return { stdout, stderr: "", exitCode: 0 }
  } catch (error: unknown) {
    const err = error as Error & { stdout?: string; stderr?: string; status?: number; code?: string }
    if (err.code === "ETIMEDOUT" || (err as { killed?: boolean }).killed) {
      return {
        stdout: "",
        stderr: `rclone command timed out after ${timeoutMs ?? 0}ms. Check your connection or try manually: ${cmd}`,
        exitCode: 124,
      }
    }
    return {
      stdout: err.stdout ?? "",
      stderr: err.stderr ?? "",
      exitCode: err.status ?? 1,
    }
  }
}

/**
 * Run an rclone operation with retries.
 * Each attempt has a 5-minute timeout. On failure, waits with exponential
 * backoff (2s, 4s, 6s) before retrying, up to `maxAttempts` total.
 */
async function rcloneWithRetry(
  args: string[],
  label: string,
  maxAttempts = 3,
): Promise<void> {
  let lastErr: Error | null = null
  for (let attempt = 1; attempt <= maxAttempts; attempt++) {
    const result = rclone(args, 5 * 60_000)
    if (result.exitCode === 0) return
    lastErr = new Error(`${label}: ${result.stderr || result.stdout}`.trim())
    if (attempt < maxAttempts) {
      await sleep(attempt * 2000) // 2s, 4s, 6s
    }
  }
  throw lastErr!
}

function checkRemote(remote: string): void {
  const result = rclone(["listremotes"], 15_000)
  if (result.exitCode !== 0) {
    throw new Error(
      `rclone is not available or config not found.\nConfig searched: ${RCLONE_CONFIG}\nInstall rclone: sudo apt install rclone\nThen run: rclone config`,
    )
  }
  const remotes = result.stdout.trim().split("\n").filter(Boolean).map((r) => r.replace(/:$/, ""))
  if (!remotes.includes(remote)) {
    throw new Error(
      `rclone remote "${remote}" not found. Configure it: rclone config\nAvailable remotes: ${remotes.join(", ") || "(none)"}`,
    )
  }
}

export class RcloneStorageBackend implements StorageBackend {
  private readonly remote: string
  private readonly basePath: string
  private checked = false

  constructor(remote: string, basePath: string) {
    this.remote = remote
    this.basePath = basePath
  }

  private ensureRemote(): void {
    if (!this.checked) {
      checkRemote(this.remote)
      this.checked = true
    }
  }

  private remotePath(path: string): string {
    const prefix = this.basePath ? `${this.basePath}/` : ""
    return `${this.remote}:${prefix}${path}`
  }

  async uploadFile(localPath: string, remotePath: string): Promise<void> {
    this.ensureRemote()
    await rcloneWithRetry(["copyto", localPath, this.remotePath(remotePath)], "rclone upload failed")
  }

  async downloadFile(remotePath: string, localPath: string): Promise<void> {
    this.ensureRemote()
    await mkdir(join(localPath, ".."), { recursive: true })
    await rcloneWithRetry(["copyto", this.remotePath(remotePath), localPath], "rclone download failed")
  }

  async exists(remotePath: string): Promise<boolean> {
    this.ensureRemote()
    const result = rclone(["lsf", this.remotePath(remotePath)], 30_000)
    if (result.exitCode !== 0) return false
    return result.stdout.trim().length > 0
  }

  async list(prefix: string): Promise<string[]> {
    this.ensureRemote()
    const result = rclone(["lsf", this.remotePath(prefix)], 30_000)
    if (result.exitCode !== 0) return []
    return result.stdout.trim().split("\n").filter(Boolean)
  }

  async remove(remotePath: string): Promise<void> {
    this.ensureRemote()
    // purge removes the directory and all its contents recursively
    const result = rclone(["purge", this.remotePath(remotePath)])
    if (result.exitCode !== 0) {
      // If it doesn't exist, that's fine
      if (result.stderr?.includes("directory not found") || result.stderr?.includes("doesn't exist")) return
      throw new Error(`rclone remove failed: ${result.stderr || result.stdout}`)
    }
  }
}

File: internal/storage/rclone.go
package storage

import (
	"bytes"
	"fmt"
	"io"
	"os/exec"
	"strings"
)

type RcloneBackend struct {
	Remote     string
	ConfigPath string
}

func NewRcloneBackend(remote string) *RcloneBackend {
	return &RcloneBackend{
		Remote:     remote,
		ConfigPath: "/home/keraki/.config/rclone/rclone.conf",
	}
}

func (r *RcloneBackend) Upload(remotePath string, in io.Reader) error {
	fullDest := fmt.Sprintf("%s/%s", r.Remote, remotePath)

	cmd := exec.Command("rclone", "rcat", "--config", r.ConfigPath, fullDest)

	cmd.Stdin = in

	var stderr bytes.Buffer
	cmd.Stderr = &stderr
	if err := cmd.Run(); err != nil {
		return fmt.Errorf("failed to upload to %s via rclone rcat: %w\nRclone Error: %s", fullDest, err, stderr.String())
	}

	return nil
}

func (r *RcloneBackend) Download(remotePath string, out io.Writer) error {
	fullSrc := fmt.Sprintf("%s/%s", r.Remote, remotePath)

	cmd := exec.Command("rclone", "cat", "--config", r.ConfigPath, fullSrc)

	cmd.Stdout = out

	var stderr bytes.Buffer
	cmd.Stderr = &stderr
	if err := cmd.Run(); err != nil {
		return fmt.Errorf("failed to download from %s via rclone cat: %w\nRclone Error: %s", fullSrc, err, stderr.String())
	}

	return nil
}

// ListFiles lists files at the given remote path prefix using rclone lsjson.
// Each prefix segment is joined with "/" to form the remote directory path.
// Returns the raw JSON output from rclone lsjson.
func (r *RcloneBackend) ListFiles(prefix []string) ([]byte, error) {
	remotePath := r.Remote
	if len(prefix) > 0 {
		remotePath = fmt.Sprintf("%s/%s", r.Remote, strings.Join(prefix, "/"))
	}

	cmd := exec.Command("rclone", "lsjson", "--config", r.ConfigPath, remotePath)

	var stdout, stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr
	if err := cmd.Run(); err != nil {
		return nil, fmt.Errorf("failed to list %s via rclone lsjson: %w\nRclone Error: %s", remotePath, err, stderr.String())
	}

	return stdout.Bytes(), nil
}

// Delete removes a file or directory at the given remote path using rclone purge.
func (r *RcloneBackend) Delete(remotePath string) error {
	fullPath := fmt.Sprintf("%s/%s", r.Remote, remotePath)

	cmd := exec.Command("rclone", "purge", "--config", r.ConfigPath, fullPath)

	var stderr bytes.Buffer
	cmd.Stderr = &stderr
	if err := cmd.Run(); err != nil {
		return fmt.Errorf("failed to delete %s via rclone purge: %w\nRclone Error: %s", fullPath, err, stderr.String())
	}

	return nil
}
